### PR TITLE
Add CORAL for aligning C_1 and \hat{C}_2

### PR DESCRIPTION
**Pull request by mdml**
### Contributions

My contributions were to `example/sl-prediction`:

1. I added model selection (using [`GridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)) to `train_and_predict.py`. We now automatically select `C` for the linear SVM and `n_estimators` (number of trees) for the random forest. Note that by default we do the grid search, so runtimes will be slower.
2. I added a `normalize_embeddings` step (and script) that occurs after HANDL and before featurization/training/prediction. There are currently two supported normalizations: `"none"` and `"coral"`.

### Testing

You can test using the toy data, which is currently set up to use CORAL normalization:

    snakemake --configfile=configs/toy.json

### CODE REVIEW

File: example/sl-prediction/normalize_embedding.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    from numpy.linalg import pinv, inv
+    C_S = np.cov(D_S.T) + lamb*np.eye(D_S.shape[1])
+    C_T = np.cov(D_T.T) + lamb*np.eye(D_T.shape[1])
+    C_S_centered = scale(D_S, with_std = False)
+    C_T_centered = scale(D_T, with_std = False)
+    
+    # Perform CORAL (Algorithm 1)
+    if source_to_target:
+        # this is the recommended approach in the coral paper
+        C_S_sqrt = np.real_if_close(sqrtm(C_S))
```
Comment (theJasonFan) [inline example/sl-prediction/normalize_embedding.py]: Hi Max, where is pipeline taking more time? FYI `sqrtm` is not parallelized so it runs slower than `rkhs_factor` implemented in `handl.py`.

```
>>> timeit.timeit('rkhs_factor(A)', setup=setup_str, number=3)
1.3135006160009652
>>> timeit.timeit('sqrtm(A)', setup=setup_str, number=3)
6.155250631971285
>>> setup_str
'import numpy as np; from scipy.linalg import sqrtm; from handl import rkhs_factor;A = np.random.rand(1000,1000);A = A @ A.T;
```

File: example/sl-prediction/normalize_embedding.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    - D_S^*: Adjusted source data
+    """
+    # Set up
+    from sklearn.preprocessing import scale
+    from scipy.linalg import sqrtm
+    from numpy.linalg import pinv, inv
+    C_S = np.cov(D_S.T) + lamb*np.eye(D_S.shape[1])
+    C_T = np.cov(D_T.T) + lamb*np.eye(D_T.shape[1])
+    C_S_centered = scale(D_S, with_std = False)
+    C_T_centered = scale(D_T, with_std = False)
```
Comment (theJasonFan) [inline example/sl-prediction/normalize_embedding.py]: Doesn't coral require both matrices to be of unit variance? If we want to not change the variance of the matrix being projected on to, we might have to think more carefully about how to scale the variances.

File: example/sl-prediction/train_and_predict.py [hunk 3]
```diff
@@ -137,8 +145,11 @@ def train_and_predict(fold, _pairs_train, _X_train, _y_train, _pairs_test, _X_te
     # SVM
     elif args.classifier == 'svm':
         # Train the Linear SVM
-        clf = LinearSVC(C=args.svm_C, tol=args.svm_tolerance)
+        svc = LinearSVC(tol=args.svm_tolerance, random_state=_random_state+1)
+        clf = GridSearchCV(svc, dict(C=args.svm_Cs), cv=inner_cv,
```
Comment (theJasonFan) [inline example/sl-prediction/train_and_predict.py:149]: You can speed this up by giving `n_jobs` parameter to GridSearchCV and parallelize the training.

**Review by theJasonFan — COMMENTED** [review]
Please see comments about CORAL.

We should call `rkhs_factor` instead of `sqrtm`, the former is much faster.

GridSearchCV can also be parallelized

**Comment by mcrovella** [comment]
Hi - no coral does not require unit variance and I don’t think we want that.   Also, sqrtm returns the “principal” square root which is the only symmetric positive definite square root.   I spent a bit of time learning about this and can update you on our call.

File: example/sl-prediction/normalize_embedding.py (hunk 1 above)
Comment (theJasonFan) [reply example/sl-prediction/normalize_embedding.py]: Saw Mark's comment, will wait for update...

**Review by theJasonFan — APPROVED** [review]
Looks good. Merging!
